### PR TITLE
Refactor microsoft_best_fit_least_utilized method for VM Placement.

### PR DESCRIPTION
**Pull request by billfitzgerald0120**
Refactored method and test.
Removed old test.

This PR is based on the issue below.
ManageIQ/manageiq#12038

@miq-bot add_label refactoring
@miq-bot assign @tinaafitz

### CODE REVIEW

File: content/automate/ManageIQ/Infrastructure/VM/Provisioning/Placement.class/__methods__/microsoft_best_fit_least_utilized.rb [hunk 1]
```diff
@@ -1,45 +1,78 @@
...
+                host = storage = min_registered_vms = nil
+                request.eligible_hosts.select { |h| !h.maintenance && h.power_state == "on" }.each do |h|
+                  next if min_registered_vms && h.vms.size >= min_registered_vms
 
-prov = $evm.root["miq_provision"]
-vm   = prov.vm_template
-raise "VM not specified" if vm.nil?
-
-ems = vm.ext_management_system
-raise "EMS not found for VM [#{vm.name}]" if ems.nil?
```
Comment (lfu) [inline content/automate/ManageIQ/Infrastructure/VM/Provisioning/Placement.class/__methods__/microsoft_best_fit_least_utilized.rb]: No need to check EMS any more?

File: content/automate/ManageIQ/Infrastructure/VM/Provisioning/Placement.class/__methods__/microsoft_best_fit_least_utilized.rb [hunk 2]
```diff
@@ -1,45 +1,78 @@
...
-    unless s.nil?
-      host    = h
-      storage = s
-      min_registered_vms = nvms
+                  # Setting the host to filter eligible storages
+                  request.set_host(h)
+
+                  # Filter out storages that do not have enough free space for the Vm
+                  storages = request.eligible_storages.find_all { |s| s.free_space > vm.provisioned_storage }
+                  storages.select! { |s| s.storage_profiles.pluck(:id).include?(storage_profile_id) } if storage_profile_id
```
Comment (lfu) [inline content/automate/ManageIQ/Infrastructure/VM/Provisioning/Placement.class/__methods__/microsoft_best_fit_least_utilized.rb]: We are adding the checking of storage profile.

File: content/automate/ManageIQ/Infrastructure/VM/Provisioning/Placement.class/__methods__/microsoft_best_fit_least_utilized.rb (hunk 1 above)
Comment (billfitzgerald0120) [reply content/automate/ManageIQ/Infrastructure/VM/Provisioning/Placement.class/__methods__/microsoft_best_fit_least_utilized.rb]: done

File: content/automate/ManageIQ/Infrastructure/VM/Provisioning/Placement.class/__methods__/microsoft_best_fit_least_utilized.rb (hunk 2 above)
Comment (billfitzgerald0120) [reply content/automate/ManageIQ/Infrastructure/VM/Provisioning/Placement.class/__methods__/microsoft_best_fit_least_utilized.rb]: I removed this, thanks

File: content/automate/ManageIQ/Infrastructure/VM/Provisioning/Placement.class/__methods__/microsoft_best_fit_least_utilized.rb [hunk 3]
```diff
@@ -1,45 +1,75 @@
...
+              end
+
+              def vm
+                vm_obj = request.vm_template.tap do |vm|
+                  raise 'VM Object not specified' if vm.nil?
+                end
+                vm_obj.ext_management_system.tap do |ems|
+                  raise 'ext_management_system not specified' if ems.nil?
+                end
+                vm_obj
```
Comment (lfu) [inline content/automate/ManageIQ/Infrastructure/VM/Provisioning/Placement.class/__methods__/microsoft_best_fit_least_utilized.rb]: How about this?
```
def vm
    request.vm_template.tap do |vm|
      raise 'VM Object not specified' if vm.nil?
      raise "EMS not found for VM [#{vm.name}]" if vm.ext_management_system.nil?
    end
end
```

Comment (billfitzgerald0120) [reply content/automate/ManageIQ/Infrastructure/VM/Provisioning/Placement.class/__methods__/microsoft_best_fit_least_utilized.rb]: That looks really good, let me change it and test everything.  Thanks

File: content/automate/ManageIQ/Infrastructure/VM/Provisioning/Placement.class/__methods__/microsoft_best_fit_least_utilized.rb (hunk 2 above)
Comment (lfu) [reply content/automate/ManageIQ/Infrastructure/VM/Provisioning/Placement.class/__methods__/microsoft_best_fit_least_utilized.rb]: We can add it in a separate PR if it is required to check against storage profile as this PR is meant for refactoring.

Comment (billfitzgerald0120) [reply content/automate/ManageIQ/Infrastructure/VM/Provisioning/Placement.class/__methods__/microsoft_best_fit_least_utilized.rb]: ok, thanks

File: content/automate/ManageIQ/Infrastructure/VM/Provisioning/Placement.class/__methods__/microsoft_best_fit_least_utilized.rb (hunk 3 above)
Comment (billfitzgerald0120) [reply content/automate/ManageIQ/Infrastructure/VM/Provisioning/Placement.class/__methods__/microsoft_best_fit_least_utilized.rb]: done

**Comment by gmcculloug** [comment]
@lfu Looks like your review comments were addressed.  Anything else on this one or is it ready to merge?

**Pull request merged**